src/staking-v3/logic/services/DappStakingService.ts

```diff
@@ -0,0 +1,684 @@
+    unstakeFromAddress: string,
+    unstakeAmount: bigint
+  ): Promise<ExtrinsicPayload> {
+    const calls: ExtrinsicPayload[] = [];
+
+    // Staker rewards
+    const claimStakerCall = await this.getClaimStakerRewardsCall(senderAddress);
+    claimStakerCall && calls.push(...claimStakerCall);
+    // Bonus rewards
+    const claimBonusCalls = await this.getClaimBonusRewardsCalls(senderAddress);
+    claimBonusCalls && calls.push(...claimBonusCalls);
+    // Lock tokens
+    const lockCall =
+      amountToLock > 0 ? await this.dappStakingRepository.getLockCall(amountToLock) : undefined;
+    lockCall && calls.push(lockCall);
+    // Cleanup expired entries if we reached maxNumberOfStakedContracts
+    if (await this.shouldCleanupExpiredEntries(senderAddress)) {
+      const cleanupCall = await this.dappStakingRepository.getCleanupExpiredEntriesCall();
+      calls.push(cleanupCall);
+    }
+    // Unstake tokens
+    if (unstakeAmount > BigInt(0)) {
+      const unstakeCall = await this.dappStakingRepository.getUnstakeCall(
+        unstakeFromAddress,
+        Number(ethers.utils.formatEther(unstakeAmount.toString()))
+      );
+      calls.push(unstakeCall);
+    }
+    // Stake tokens
+    for (const info of stakeInfo) {
+      calls.push(await this.dappStakingRepository.getStakeCall(info.address, info.amount));
+    }
+
+    const batch = await this.dappStakingRepository.batchAllCalls(calls);
+
+    return batch;
+  }
+
+  private async shouldCleanupExpiredEntries(senderAddress: string): Promise<boolean> {
+    const [stakerInfo, constants, protocolState] = await Promise.all([
+      this.dappStakingRepository.getStakerInfo(senderAddress, true),
+      this.dappStakingRepository.getConstants(),
+      this.dappStakingRepository.getProtocolState(),
+    ]);
+
+    let expiredEntries = 0;
+
+    stakerInfo.forEach((info) => {
+      if (
+        (info.staked.period < protocolState.periodInfo.number && !info.loyalStaker) ||
+        info.staked.period < protocolState.periodInfo.number - constants.rewardRetentionInPeriods
+      ) {
+        expiredEntries++;
+      }
+    });
+
+    return expiredEntries > 0;
+  }
+
+  private async getBonusRewardsAndContractsToClaim(
+    senderAddress: string
+  ): Promise<{ rewards: bigint; contractsToClaim: string[] }> {
+    let result = { rewards: BigInt(0), contractsToClaim: Array<string>() };
+    const [stakerInfo, protocolState, constants] = await Promise.all([
+      this.dappStakingRepository.getStakerInfo(senderAddress, true),
+      this.dappStakingRepository.getProtocolState(),
+      this.dappStakingRepository.getConstants(),
+    ]);
+
+    const multiplier = await this.getMultiplier();
+    for (const [contract, info] of stakerInfo.entries()) {
+      // Staker is eligible to bonus rewards if he is a loyal staker and if rewards are not expired
+      // and if stake amount refers to the past period.
+      if (
+        info.loyalStaker &&
+        protocolState &&
```

Comment:
Just out of interest, when can this be `null`?

Reply:
This can happen in very rare case of communication error.